### PR TITLE
Add flag to specify TLS minimum version.

### PR DESCRIPTION
**Pull request by fische**
This adds a new `--tls-min-version` flag to allow passing the minimum accepted version of TLS for new connections. It supports the following values: `1.0`, `1.1`, `1.2` and  `1.3`.
Without the flag:
```
$ plz-out/bin/elan/elan --port 7777 -k ./key.pem -c ./cert.pem -s file://$(mktemp -d) --parallelism 20 --dir_cache_size 40960
$ nmap --script ssl-enum-ciphers -p 7777 localhost

Starting Nmap 7.60 ( https://nmap.org ) at 2021-11-09 13:53 GMT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000055s latency).
Other addresses for localhost (not scanned): ::1

PORT     STATE SERVICE
7777/tcp open  cbt
| ssl-enum-ciphers: 
|   TLSv1.0: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|     compressors: 
|       NULL
|     cipher preference: server
|     warnings: 
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|   TLSv1.1: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|     compressors: 
|       NULL
|     cipher preference: server
|     warnings: 
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|     compressors: 
|       NULL
|     cipher preference: server
|     warnings: 
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|_  least strength: C

Nmap done: 1 IP address (1 host up) scanned in 0.25 seconds
```

With the flag set to TLS 1.3:
```
$ plz-out/bin/elan/elan --port 7777 -k ./key.pem -c ./cert.pem -s file://$(mktemp -d) --parallelism 20 --dir_cache_size 40960 --tls-min-version 1.3
$ nmap --script ssl-enum-ciphers -p 7777 localhost

Starting Nmap 7.60 ( https://nmap.org ) at 2021-11-09 13:52 GMT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000054s latency).
Other addresses for localhost (not scanned): ::1

PORT     STATE SERVICE
7777/tcp open  cbt

Nmap done: 1 IP address (1 host up) scanned in 0.13 seconds
```

### CODE REVIEW

File: grpcutil/interceptors.go [hunk 1]
```diff
@@ -25,13 +25,14 @@ import (
...
-	KeyFile   string `short:"k" long:"key_file" description:"Key file to load TLS credentials from"`
-	CertFile  string `short:"c" long:"cert_file" description:"Cert file to load TLS credentials from"`
-	TokenFile string `long:"token_file" description:"File containing a pre-shared token that clients must provide as authentication."`
-	AuthAll   bool   `long:"auth_all" description:"Require authentication on all RPCs (by default only on RPCs that mutate state)"`
-	NoHealth  bool   `no-flag:"true" description:"Used internally to indicate when we don't want to automatically add a healthcheck."`
+	Host          string `long:"host" description:"Host to listen on"`
+	Port          int    `short:"p" long:"port" default:"7777" description:"Port to serve on"`
+	KeyFile       string `short:"k" long:"key_file" description:"Key file to load TLS credentials from"`
+	CertFile      string `short:"c" long:"cert_file" description:"Cert file to load TLS credentials from"`
+	TLSMinVersion string `long:"tls-min-version" description:"TLS minimum version allowed on new connections"`
```
Comment (peterebden) [inline grpcutil/interceptors.go:32]: FYI you can add `choice:"1.2" choice:"1.3"` to limit these.

File: grpcutil/creds.go [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
 import (
+	"crypto/tls"
+
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
 )
 
+var tlsVersions = map[string]uint16{
+	"1.0": tls.VersionTLS10,
+	"1.1": tls.VersionTLS11,
```
Comment (peterebden) [inline grpcutil/creds.go:12]: should we even allow anyone to choose 1.0 or 1.1? I doubt we have any use case for it (I'm not sure if we even have any need to support anything below 1.3?)

**Comment by fische** [comment]
The flag defaults now to TLS 1.3:
```
$ plz-out/bin/elan/elan --port 7777 -k ./key.pem -c ./cert.pem -s file://$(mktemp -d) --parallelism 20 --dir_cache_size 40960
$ nmap --script ssl-enum-ciphers -p 7777 localhost

Starting Nmap 7.60 ( https://nmap.org ) at 2021-11-09 13:52 GMT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000054s latency).
Other addresses for localhost (not scanned): ::1

PORT     STATE SERVICE
7777/tcp open  cbt

Nmap done: 1 IP address (1 host up) scanned in 0.13 seconds
```

**Review by peterebden — APPROVED** [review]
Thanks!

YOu probably want to bump the VERSION file too to get a new release?

**Comment by fische** [comment]
> Thanks!
> 
> YOu probably want to bump the VERSION file too to get a new release?

Oh right, yes, good idea

**Pull request merged**